### PR TITLE
Tutorial waypoint follower

### PR DESCRIPTION
**Pull request by LucasBuckland**
Still need final graphics for parrot, but wanted to get a preliminary pull request in to review the code, which is ready to go.

### CODE REVIEW

File: parrot_controller/include/gl_headers.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * @file gl_headers.h
+ * @brief OpenGL headers.
+ *
+ */
+
+
+
+
+#ifndef GL_HEADERS_H
+#define	GL_HEADERS_H
```
Comment (ltbj) [inline parrot_controller/include/gl_headers.h:36]: Odd spacing.

File: parrot_controller/include/grid.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * @file grid.h
+ * @brief Grid Rendering Interface.
+ *
+ */
+
+
+
+
+#ifndef GRID_H
+#define	GRID_H
```
Comment (ltbj) [inline parrot_controller/include/grid.h:36]: Odd spacing.

File: parrot_controller/include/gui.h [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+ * @file gui.h
+ * @brief GUI Interface.
+ *
+ */
+
+
+
+
+#ifndef GUI_H
+#define	GUI_H
```
Comment (ltbj) [inline parrot_controller/include/gui.h:36]: Odd spacing.

File: parrot_controller/include/ps_interface.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ * @file ps_interface.h
+ * @brief PolySync Interface.
+ *
+ */
+
+
+
+
+#ifndef PS_INTERFACE_H
+#define	PS_INTERFACE_H
```
Comment (ltbj) [inline parrot_controller/include/ps_interface.h:36]: Odd spacing.

File: parrot_controller/include/render.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * @file render.h
+ * @brief Primitive Rendering Interface.
+ *
+ */
+
+
+
+
+#ifndef RENDER_H
+#define	RENDER_H
```
Comment (ltbj) [inline parrot_controller/include/render.h:36]: Odd spacing.

File: parrot_controller/src/ps_interface.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    node_data = (node_data_s*) user_data;
+
+    // ignore if not valid
+    if( (node_data == NULL) || (message == NULL) )
+    {
+        return;
+    }
+
+    // get source GUID
+    if( psync_message_get_source_guid( message, &src_guid ) != DTC_NONE )
```
Comment (ltbj) [inline parrot_controller/src/ps_interface.c]: Are you storing this for future use?

File: parrot_controller/src/vehicle_control.c [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+    }
+}
+
+
+int calculate_steering_angle_based_on_goal_waypoint( 
+        waypoint_s * goalWaypoint, 
+        waypoint_s currentPosition,
+        double * steeringAngle )
+{
+    if( ! goalWaypoint->valid )
```
Comment (ltbj) [inline parrot_controller/src/vehicle_control.c]: Can we have explicit checks against NULL instead, to follow style guidelines?

**Review by ltbj — COMMENTED** [review]
I made it to the parrot visualizer Makefile

File: parrot_controller/Makefile [hunk 8]
```diff
@@ -0,0 +1,89 @@
+##########################################################
+# makefile for viewer-lite
```
Comment (jlamb-at-polysync) [inline parrot_controller/Makefile]: lot of old viewer lite stuff...

File: parrot_controller/Makefile [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+all: dirs $(TARGET)
+
+#
+ifneq ($(XDEPS),)
+include $(XDEPS)
+endif
+
+# directories
+dirs::
+	mkdir -p doc/html
```
Comment (jlamb-at-polysync) [inline parrot_controller/Makefile]: is this doc stuff still in use?

File: parrot_controller/include/vehicle_control.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+
+#include "math.h"
+
+#include "gui.h"
+
+
+
+
+#define RAD2DEG (180.0/M_PI)
```
Comment (jlamb-at-polysync) [inline parrot_controller/include/vehicle_control.h]: aren't these already defined in the common header?

File: parrot_controller/include/gl_headers.h (hunk 1 above)
Comment (LucasBuckland) [reply parrot_controller/include/gl_headers.h:36]: This doesn't appear in my local files, not sure why Github is showing it like this.

File: parrot_controller/src/ps_interface.c (hunk 6 above)
Comment (LucasBuckland) [reply parrot_controller/src/ps_interface.c]: Removed.

File: parrot_controller/src/vehicle_control.c (hunk 7 above)
Comment (LucasBuckland) [reply parrot_controller/src/vehicle_control.c]: Fixed.

File: parrot_controller/Makefile (hunk 8 above)
Comment (LucasBuckland) [reply parrot_controller/Makefile]: Fixed.

File: parrot_controller/Makefile (hunk 9 above)
Comment (LucasBuckland) [reply parrot_controller/Makefile]: Removed.

File: parrot_controller/include/vehicle_control.h (hunk 10 above)
Comment (LucasBuckland) [reply parrot_controller/include/vehicle_control.h]: Deleted and switched to using common.

File: parrot_controller/include/gl_headers.h (hunk 1 above)
Comment (jlamb-at-polysync) [reply parrot_controller/include/gl_headers.h:36]: probably a tab showing up as a space?
try replacing it with a space character maybe

File: parrot_controller/include/vehicle_control.h (hunk 10 above)
Comment (jlamb-at-polysync) [inline parrot_controller/include/vehicle_control.h]: Can you add code documentation for all of these macro's?

File: parrot_controller/include/vehicle_control.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+#define DEG2RAD (M_PI/180.0)
+
+#define STEERING_CONTROL_FACTOR 2000
```
Comment (jlamb-at-polysync) [inline parrot_controller/include/vehicle_control.h:48]: I think our style guide has this, but we use parentheses when possible on macro values

File: parrot_controller/include/waypoint.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * @brief Waypoint.
+ *
+ */
+typedef struct
+{
+    //
+    //
+    double                      x; /*!< Waypoint X coordinate. */
```
Comment (jlamb-at-polysync) [inline parrot_controller/include/waypoint.h:59]: our c-code style doesn't tab/indent things any more, can you un-tab these?

File: parrot_controller/include/waypoint.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    //
+    double                      heading; /*!< Waypoint heading. */
+    //
+    //
+    int                         valid; /*!< Is Waypoint valid */
+} waypoint_s;
+
+
+
+void zero_waypoints( waypoint_s * waypoints );
```
Comment (jlamb-at-polysync) [inline parrot_controller/include/waypoint.h:73]: c-style calls for constant pointer when possible, same for pointer to constant

also documentation is missing

File: parrot_controller/include/waypoint.h [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+void add_new_waypoint( double x, double y, waypoint_s * waypoints );
```
Comment (jlamb-at-polysync) [inline parrot_controller/include/waypoint.h:75]: c-style calls for constant pointer when possible, same for pointer to constant, and constant data

also documentation is missing

**Review by ltbj — APPROVED** [review]

**Pull request closed without merging**